File: src/api/checksums.ts
// ui/src/api/checksums.ts
//
// Tauri API wrappers for checksum calculation functions.
// These call the Rust backend for checksum calculations.

import { invoke } from "@tauri-apps/api/core";

// ============================================================================
// Types
// ============================================================================

/**
 * Supported checksum algorithms.
 * Must match the Rust ChecksumAlgorithm enum.
 */
export type ChecksumAlgorithm =
  | "xor"
  | "sum8"
  | "crc8"
  | "crc8_sae_j1850"
  | "crc8_autosar"
  | "crc8_maxim"
  | "crc8_cdma2000"
  | "crc8_dvb_s2"
  | "crc8_nissan"
  | "crc16_modbus"
  | "crc16_ccitt";

/**
 * Result of checksum validation.
 */
export interface ChecksumValidationResult {
  /** The checksum value extracted from the frame */
  extracted: number;
  /** The calculated checksum value */
  calculated: number;
  /** Whether the checksum is valid (extracted === calculated) */
  valid: boolean;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Calculate checksum using the specified algorithm with byte range.
 *
 * @param algorithm - The checksum algorithm to use
 * @param data - The complete frame data as bytes
 * @param calcStartByte - First byte index to include (supports negative indexing)
 * @param calcEndByte - Last byte index exclusive (supports negative indexing)
 * @returns The calculated checksum value
 */
export async function calculateChecksum(
  algorithm: ChecksumAlgorithm,
  data: number[] | Uint8Array,
  calcStartByte: number,
  calcEndByte: number
): Promise<number> {
  const dataArray = data instanceof Uint8Array ? Array.from(data) : data;
  return invoke<number>("calculate_checksum_cmd", {
    algorithm,
    data: dataArray,
    calcStartByte,
    calcEndByte,
  });
}

/**
 * Validate a checksum in frame data.
 *
 * @param algorithm - The checksum algorithm to use
 * @param data - The complete frame data as bytes
 * @param startByte - Byte offset where checksum is stored (supports negative indexing)
 * @param byteLength - Length of checksum (1 or 2 bytes)
 * @param bigEndian - true for big-endian, false for little-endian
 * @param calcStartByte - First byte to include in calculation (supports negative indexing)
 * @param calcEndByte - Last byte (exclusive) to include (supports negative indexing)
 * @returns Validation result with extracted, calculated, and valid fields
 */
export async function validateChecksum(
  algorithm: ChecksumAlgorithm,
  data: number[] | Uint8Array,
  startByte: number,
  byteLength: number,
  bigEndian: boolean,
  calcStartByte: number,
  calcEndByte: number
): Promise<ChecksumValidationResult> {
  const dataArray = data instanceof Uint8Array ? Array.from(data) : data;
  return invoke<ChecksumValidationResult>("validate_checksum_cmd", {
    algorithm,
    data: dataArray,
    startByte,
    byteLength,
    bigEndian,
    calcStartByte,
    calcEndByte,
  });
}

/**
 * Resolve a byte index, supporting negative indexing.
 * Negative indices count from the end: -1 = last byte, -2 = second-to-last, etc.
 *
 * @param index - The byte index (can be negative)
 * @param frameLength - Total frame length in bytes
 * @returns The resolved absolute byte index
 */
export async function resolveByteIndex(
  index: number,
  frameLength: number
): Promise<number> {
  return invoke<number>("resolve_byte_index_cmd", {
    index,
    frameLength,
  });
}

// ============================================================================
// Parameterised CRC Functions (for Discovery)
// ============================================================================

/**
 * CRC parameters for parameterised calculation.
 */
export interface CrcParameters {
  polynomial: number;
  init: number;
  xorOut: number;
  reflect: boolean;
}

/**
 * Result of batch CRC testing.
 */
export interface BatchDiscoveryResult {
  matchCount: number;
  totalCount: number;
}

/**
 * Calculate CRC-8 with arbitrary parameters.
 *
 * @param data - The data to calculate CRC over
 * @param polynomial - The CRC polynomial (0x00-0xFF)
 * @param init - Initial CRC value
 * @param xorOut - Final XOR value
 * @param reflect - Whether to use reflected (LSB-first) mode
 * @returns The calculated CRC-8 value
 */
export async function crc8Parameterised(
  data: number[] | Uint8Array,
  polynomial: number,
  init: number,
  xorOut: number,
  reflect: boolean
): Promise<number> {
  const dataArray = data instanceof Uint8Array ? Array.from(data) : data;
  return invoke<number>("crc8_parameterised_cmd", {
    data: dataArray,
    polynomial,
    init,
    xorOut,
    reflect,
  });
}

/**
 * Calculate CRC-16 with arbitrary parameters.
 *
 * @param data - The data to calculate CRC over
 * @param polynomial - The CRC polynomial (0x0000-0xFFFF)
 * @param init - Initial CRC value
 * @param xorOut - Final XOR value
 * @param reflectIn - Whether to reflect input bytes
 * @param reflectOut - Whether to reflect the final CRC output
 * @returns The calculated CRC-16 value
 */
export async function crc16Parameterised(
  data: number[] | Uint8Array,
  polynomial: number,
  init: number,
  xorOut: number,
  reflectIn: boolean,
  reflectOut: boolean
): Promise<number> {
  const dataArray = data instanceof Uint8Array ? Array.from(data) : data;
  return invoke<number>("crc16_parameterised_cmd", {
    data: dataArray,
    polynomial,
    init,
    xorOut,
    reflectIn,
    reflectOut,
  });
}

/**
 * Batch test a CRC configuration against multiple payloads.
 * Optimised for checksum discovery - tests one polynomial/config
 * against many frames in a single IPC call.
 *
 * @param payloads - Array of frame payloads to test
 * @param expectedChecksums - Expected checksum values for each payload
 * @param checksumBits - 8 for CRC-8, 16 for CRC-16
 * @param polynomial - The CRC polynomial to test
 * @param init - Initial CRC value
 * @param xorOut - Final XOR value
 * @param reflect - Whether to use reflected mode
 * @returns Match statistics
 */
export async function batchTestCrc(
  payloads: number[][],
  expectedChecksums: number[],
  checksumBits: 8 | 16,
  polynomial: number,
  init: number,
  xorOut: number,
  reflect: boolean
): Promise<BatchDiscoveryResult> {
  const result = await invoke<{ match_count: number; total_count: number }>(
    "batch_test_crc_cmd",
    {
      payloads,
      expectedChecksums,
      checksumBits,
      polynomial,
      init,
      xorOut,
      reflect,
    }
  );
  return {
    matchCount: result.match_count,
    totalCount: result.total_count,
  };
}

/** One point in the checksum candidate space. */
export interface ChecksumSpec {
  algorithm: ChecksumAlgorithm;
  /** Byte offset of the checksum; negative counts from the end. */
  position: number;
  byteLength: 1 | 2;
  bigEndian: boolean;
  calcStartByte: number;
  calcEndByte: number;
}

export interface ChecksumSpecResult {
  /** Index into the `specs` array passed in. */
  specIndex: number;
  matchCount: number;
  /** Frames the spec fitted — frames too short for it are excluded, not counted as misses. */
  totalCount: number;
}

export interface ChecksumSweepResponse {
  results: ChecksumSpecResult[];
}

/**
 * A translatable note. Rust decides *what* to say, the frontend decides how —
 * render with `t(\`serial.checksumNote.${code}\`, values)`.
 */
export interface ChecksumNote {
  code: string;
  values: Record<string, string | number>;
}

/**
 * What one end-relative byte column looks like across the sample: the structural
 * evidence behind the detector's priors.
 */
export interface ChecksumColumnStat {
  /** Negative index, e.g. -1 for the last byte. */
  position: number;
  distinctValues: number;
  /** Set when the column holds one value across every sampled frame. */
  constantValue: number | null;
  sampleCount: number;
}

export interface ChecksumCandidate {
  algorithm: ChecksumAlgorithm;
  position: number;
  length: 1 | 2;
  bigEndian: boolean;
  calcStartByte: number;
  calcEndByte: number;
  matchCount: number;
  totalCount: number;
  /** 0-100 */
  matchRate: number;
  /** 0-100 composite score. */
  confidence: number;
  notes: ChecksumNote[];
  /** Other calculation ranges that scored identically. */
  equivalentRanges: { calcStartByte: number; calcEndByte: number }[];
}

export interface ChecksumDetectionResult {
  candidates: ChecksumCandidate[];
  bestCandidate: ChecksumCandidate | null;
  tailColumns: ChecksumColumnStat[];
  /** Result-level explanation, including why nothing was found. */
  notes: ChecksumNote[];
}

export interface ChecksumDetectionOptions {
  /** Checksum offsets to try, end-relative (default [-1, -2, -3]). */
  positions?: number[];
  /** Restrict to checksums of these byte lengths; omit for both. */
  lengths?: (1 | 2)[];
  /**
   * Byte offsets just past a declared header field, from the view's ID/Source
   * chips. These widen the calculation-range candidates; they never narrow them.
   */
  headerBoundaries?: number[];
  /** Percentage below which a candidate is discarded (default 50). */
  minMatchRate?: number;
  /** Confidence below which a candidate is discarded (default 35). */
  minConfidence?: number;
}

/**
 * Rank the checksum configurations that explain a set of frames.
 *
 * The whole engine — candidate space, structural priors, confidence scoring —
 * lives in Rust beside the algorithms, so there is one implementation of each.
 * One IPC call for the whole search.
 */
export async function detectChecksum(
  frames: number[][],
  options: ChecksumDetectionOptions = {}
): Promise<ChecksumDetectionResult> {
  return invoke<ChecksumDetectionResult>("detect_checksum_cmd", { frames, options });
}

/**
 * Check specific checksum configurations against frames.
 *
 * For the live match rate behind a hand-edited configuration, where the caller
 * has one spec rather than a space to search. Unlike `batchTestCrc` above, the
 * response needs no field mapping — the Rust structs are `rename_all = "camelCase"`.
 */
export async function sweepChecksumSpecs(
  frames: number[][],
  specs: ChecksumSpec[]
): Promise<ChecksumSweepResponse> {
  if (specs.length === 0) return { results: [] };
  return invoke<ChecksumSweepResponse>("sweep_checksum_specs_cmd", { frames, specs });
}

File: src-tauri/src/checksums.rs
// ui/src-tauri/src/checksums.rs
//
// The frontend's door to `wiretap-checksum`. The algorithms, the candidate
// sweep, the scoring and the notes all live in the crate — a catalogue's
// checksum config is a catalogue concern, and keeping one implementation is
// what stopped the serial dialog and the Serial Payload tool disagreeing about
// the same bytes. What is left here is the Tauri surface: argument shapes,
// wire types, and the string-to-enum parse the frontend needs.

use serde::{Deserialize, Serialize};
use std::str::FromStr;

use wiretap_checksum::{
    calculate_checksum, crc16_parameterised, crc8_parameterised, detect_checksum,
    resolve_byte_index, sweep_specs, validate_checksum, ChecksumAlgorithm,
    ChecksumDetectionOptions, ChecksumDetectionResult, ChecksumSpec, ChecksumSpecResult,
    ChecksumValidationResult,
};

/// Result of batch checksum discovery.
///
/// Not `camelCase` like the rest, and hand-mapped in `api/checksums.ts` because
/// of it. Retired along with `batch_test_crc_cmd` once the frontend moves to the
/// crate's solver, which answers the same question without the round trip.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BatchDiscoveryResult {
    /// Number of frames that matched
    pub match_count: usize,
    /// Total number of frames tested
    pub total_count: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChecksumSweepResponse {
    pub results: Vec<ChecksumSpecResult>,
}

/// Calculate a checksum over a byte range.
///
/// `algorithm` is a catalogue id ("xor", "sum8", "crc16_modbus", …); offsets
/// support negative indexing.
#[tauri::command]
pub fn calculate_checksum_cmd(
    algorithm: String,
    data: Vec<u8>,
    calc_start_byte: i32,
    calc_end_byte: i32,
) -> Result<u16, String> {
    let algo = ChecksumAlgorithm::from_str(&algorithm)?;
    Ok(calculate_checksum(algo, &data, calc_start_byte, calc_end_byte))
}

/// Check a frame's stored checksum against a freshly calculated one.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn validate_checksum_cmd(
    algorithm: String,
    data: Vec<u8>,
    start_byte: i32,
    byte_length: usize,
    big_endian: bool,
    calc_start_byte: i32,
    calc_end_byte: i32,
) -> Result<ChecksumValidationResult, String> {
    let algo = ChecksumAlgorithm::from_str(&algorithm)?;
    Ok(validate_checksum(
        algo,
        &data,
        start_byte,
        byte_length,
        big_endian,
        calc_start_byte,
        calc_end_byte,
    ))
}

/// Resolve a byte index, supporting negative indexing (-1 = last byte).
#[tauri::command]
pub fn resolve_byte_index_cmd(index: i32, frame_length: usize) -> usize {
    resolve_byte_index(index, frame_length)
}

/// Calculate CRC-8 with arbitrary parameters.
#[tauri::command]
pub fn crc8_parameterised_cmd(
    data: Vec<u8>,
    polynomial: u8,
    init: u8,
    xor_out: u8,
    reflect: bool,
) -> u8 {
    crc8_parameterised(&data, polynomial, init, xor_out, reflect)
}

/// Calculate CRC-16 with arbitrary parameters.
#[tauri::command]
pub fn crc16_parameterised_cmd(
    data: Vec<u8>,
    polynomial: u16,
    init: u16,
    xor_out: u16,
    reflect_in: bool,
    reflect_out: bool,
) -> u16 {
    crc16_parameterised(&data, polynomial, init, xor_out, reflect_in, reflect_out)
}

/// Test one CRC configuration against many payloads.
///
/// The unit the frontend's polynomial brute force iterates — one IPC call per
/// polynomial, which is why an exhaustive CRC-16 search never finishes. The
/// crate's `solve_crc` replaces the whole loop with a single call; this stays
/// only until that lands.
#[tauri::command]
pub fn batch_test_crc_cmd(
    payloads: Vec<Vec<u8>>,
    expected_checksums: Vec<u16>,
    checksum_bits: u8,
    polynomial: u16,
    init: u16,
    xor_out: u16,
    reflect: bool,
) -> BatchDiscoveryResult {
    let total_count = payloads.len().min(expected_checksums.len());
    let match_count = payloads
        .iter()
        .zip(&expected_checksums)
        .filter(|(payload, expected)| {
            let calculated = if checksum_bits == 8 {
                crc8_parameterised(payload, polynomial as u8, init as u8, xor_out as u8, reflect)
                    as u16
            } else {
                crc16_parameterised(payload, polynomial, init, xor_out, reflect, reflect)
            };
            calculated == **expected
        })
        .count();

    BatchDiscoveryResult {
        match_count,
        total_count,
    }
}

/// Rank the checksum configurations that explain a set of frames.
#[tauri::command]
pub fn detect_checksum_cmd(
    frames: Vec<Vec<u8>>,
    options: Option<ChecksumDetectionOptions>,
) -> ChecksumDetectionResult {
    detect_checksum(&frames, &options.unwrap_or_default())
}

/// Check specific checksum configurations against frames.
///
/// Used for the live match rate behind a hand-edited configuration, where the
/// caller has one spec rather than a space to search.
#[tauri::command]
pub fn sweep_checksum_specs_cmd(
    frames: Vec<Vec<u8>>,
    specs: Vec<ChecksumSpec>,
) -> ChecksumSweepResponse {
    ChecksumSweepResponse {
        results: sweep_specs(&frames, &specs),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeSet;
    use wiretap_checksum::ALL_NOTES;

    /// Rust decides *what* the checksum detector says; the frontend decides how.
    ///
    /// Notes cross as `{ code, values }` so the prose stays translatable, which
    /// means the two halves are joined by a string rather than by a type. A code
    /// with no key renders as the raw key to the user, and a renamed
    /// interpolation variable silently drops a value — neither is a compile
    /// error.
    ///
    /// This used to be a vitest suite parsing `checksums.rs?raw`. The engine now
    /// lives in another repository where Vite cannot reach it, so the pin moved
    /// here, where both the crate's `ALL_NOTES` manifest and the locale file are
    /// in scope.
    mod note_coverage {
        use super::*;

        fn translations() -> serde_json::Map<String, serde_json::Value> {
            let locale: serde_json::Value =
                serde_json::from_str(include_str!("../../src/locales/en-AU/discovery.json"))
                    .expect("discovery.json parses");
            locale["serial"]["checksumNote"]
                .as_object()
                .expect("serial.checksumNote is an object")
                .clone()
        }

        /// `{{name}}` placeholders in a translation string.
        fn placeholders(text: &str) -> BTreeSet<String> {
            text.split("{{")
                .skip(1)
                .filter_map(|rest| rest.split_once("}}"))
                .map(|(name, _)| name.trim().to_string())
                .collect()
        }

        #[test]
        fn every_code_the_engine_emits_has_a_translation() {
            let notes = translations();
            let missing: Vec<&str> = ALL_NOTES
                .iter()
                .map(|n| n.code)
                .filter(|code| !notes.contains_key(*code))
                .collect();

            assert_eq!(missing, Vec::<&str>::new());
        }

        #[test]
        fn no_translation_exists_for_a_code_nothing_emits() {
            let notes = translations();
            let orphaned: Vec<&String> = notes
                .keys()
                .filter(|key| !ALL_NOTES.iter().any(|n| n.code == key.as_str()))
                .collect();

            assert_eq!(orphaned, Vec::<&String>::new());
        }

        #[test]
        fn translations_interpolate_exactly_the_values_the_engine_sends() {
            let notes = translations();
            let mismatched: Vec<String> = ALL_NOTES
                .iter()
                .filter_map(|note| {
                    let text = notes.get(note.code)?.as_str()?;
                    let used = placeholders(text);
                    let sent: BTreeSet<String> =
                        note.values.iter().map(|v| (*v).to_string()).collect();

                    (used != sent).then(|| {
                        format!(
                            "{}: engine sends {sent:?}, translation uses {used:?}",
                            note.code
                        )
                    })
                })
                .collect();

            assert_eq!(mismatched, Vec::<String>::new());
        }
    }

    /// The commands are thin, but the two wire shapes they own are not the
    /// crate's, so they are worth a smoke test.
    #[test]
    fn batch_test_crc_counts_only_the_matching_payloads() {
        let payloads = vec![vec![0x01u8, 0x02, 0x03], vec![0x04, 0x05, 0x06]];
        let expected = vec![crc8_parameterised(&payloads[0], 0x07, 0, 0, false) as u16, 0xFF];

        let result = batch_test_crc_cmd(payloads, expected, 8, 0x07, 0, 0, false);
        assert_eq!(result.match_count, 1);
        assert_eq!(result.total_count, 2);
    }

    #[test]
    fn an_unknown_algorithm_name_is_an_error_not_a_default() {
        assert!(calculate_checksum_cmd("nope".into(), vec![1, 2], 0, -1).is_err());
        assert_eq!(
            calculate_checksum_cmd("sum8".into(), vec![1, 2, 0], 0, -1),
            Ok(0x03)
        );
    }
}
